gnovm/pkg/gnolang/static_analysis.go

```diff
@@ -0,0 +1,268 @@
+			s.errs = append(s.errs, errors.New(fmt.Sprintf("lambda at %v does not terminate\n", ctx.f.Loc)))
+		}
+		return false, false
+	case *NameExpr:
+		return false, false
+	}
+	return false, false
+}
+
+// staticAnalysisStmt returns a boolean value,
+// indicating weather a statement is terminating or not
+func (s *StaticAnalysis) staticAnalysisStmt(stmt Stmt) bool {
+	switch n := stmt.(type) {
+	case *BranchStmt:
+		switch n.Op {
+		case BREAK:
+			ctx := s.findCtxByLabel(string(n.Label))
+			ctx.pushBreak(n)
+		case CONTINUE:
+			//
+		case DEFAULT:
+			//
+		case FALLTHROUGH:
+			return true
+		}
+	case *IfStmt:
+		terminates := s.staticAnalysisBlockStmt(n.Then.Body)
+
+		var elseTerminates bool
+		if len(n.Else.Body) > 0 {
+			elseTerminates = s.staticAnalysisBlockStmt(n.Else.Body)
+		}
+
+		return terminates && elseTerminates
+	case *ForStmt:
+		s.pushContext(&ForContext{forstmt: n})
+		_ = s.staticAnalysisBlockStmt(n.Body)
+
+		ctx := s.popContext().(*ForContext)
+
+		// there are no "break" statements referring to the "for" statement, and
+		hasNoBreaks := len(ctx.breakstmts) == 0
+		// the loop condition is absent, and
+		hasNoCond := n.Cond == nil
+
+		// the "for" statement does not use a range clause.
+		// this one is always false because in our nodes
+		// the range loop is a different data structure
+		hasRange := false
+
+		terminates := hasNoBreaks && hasNoCond && !hasRange
+
+		if !terminates {
+			return false
+		}
+
+		return true
+	// for statement
+	case *ReturnStmt:
+		// n.Results
+		return true
+	case *AssignStmt:
+		for _, rh := range n.Rhs {
+			term, is := s.staticAnalysisExpr(rh)
+
+			if is && !term {
+				return true
+			}
+		}
+		return false
+	case *SwitchStmt:
+		// there is a default case, and
+		var hasDefault bool
+		for _, clause := range n.Clauses {
+			// nil case means default
+			if clause.Cases == nil {
+				hasDefault = true
+				break
+			}
+		}
+
+		s.pushContext(&SwitchContext{switchStmt: n})
+
+		// the statement lists in each case,
+		// including the default
+		// end in a terminating statement,
+		// or a possibly labeled "fallthrough" statement.
+		casesTerm := true
+
+		for _, clause := range n.Clauses {
+			ct := s.staticAnalysisBlockStmt(clause.Body)
+			casesTerm = casesTerm && ct
+		}
+
+		ctx := s.popContext().(*SwitchContext)
+		// there are no "break" statements referring to the "switch" statement
+		hasNoBreaks := len(ctx.breakstmts) == 0
+
+		terminates := hasNoBreaks && hasDefault && casesTerm
+
+		if !terminates {
+			return false
+		}
+
+		return true
+	case *PanicStmt:
+		return true
+	}
+	return false
+}
+
+type FuncContext interface {
+	isLastExprRet() bool
+}
+
+type FuncLitContext struct {
+	hasRet bool
+	f      *FuncLitExpr
+}
+
+func (fdc *FuncLitContext) isLastExprRet() bool {
+	return fdc.hasRet
+}
+
+type FuncDeclContext struct {
+	hasRet bool
+	f      *FuncDecl
+}
+
+func (fdc *FuncDeclContext) isLastExprRet() bool {
+	return fdc.hasRet
+}
+
+type Context interface {
```

Comment:
I think you can completely drop this interface, along with `FuncContext`, and just use the actual structs straight up